src/sdc11073/sdcdevice/subscriptionmgr_base.py

```diff
@@ -0,0 +1,498 @@
+from __future__ import annotations
+
+import http.client
+import socket
+import time
+import traceback
+import uuid
+from collections import deque
+from typing import TYPE_CHECKING, List, Union, Any
+from urllib.parse import urlparse
+
+from lxml import etree as etree_
+
+from sdc11073.xml_types.addressing_types import HeaderInformationBlock
+from .. import loghelper
+from .. import multikey
+from .. import observableproperties
+from ..etc import apply_map
+from ..pysoap.soapclient import HTTPReturnCodeError
+from ..pysoap.soapenvelope import Fault, faultcodeEnum
+from ..xml_types import eventing_types as evt_types, isoduration
+from ..xml_types.basetypes import MessageType
+from ..xml_types.dpws_types import DeviceEventingFilterDialectURI
+
+if TYPE_CHECKING:
+    from ..definitions_base import BaseDefinitions
+    from ..pysoap.msgfactory import MessageFactory, CreatedMessage
+    from ..dispatch import RequestData
+    from ..pysoap.soapclientpool import SoapClientPool
+    from urllib.parse import SplitResult
+
+MAX_ROUNDTRIP_VALUES = 20
+
+
+class _RoundTripData:
+    def __init__(self, values, abs_max):
+        if values:
+            self.values = list(values)  # make a copy
+            self.min = min(values)
+            self.max = max(values)
+            self.avg = sum(values) / len(values)
+            self.abs_max = abs_max
+        else:
+            self.values = None
+            self.min = None
+            self.max = None
+            self.avg = None
+            self.abs_max = None
+
+    def __repr__(self):
+        return f'min={self.min:.4f} max={self.max:.4f} avg={self.avg:.4f} absmax={self.abs_max:.4f}'
+
+
+def _mk_dispatch_identifier(reference_parameters: list, path_suffix: str):
+    # this is always our own reference parameter. We know that is has max. one element,
+    # and the text is the identifier of the subscription
+    if path_suffix == '':
+        path_suffix = None
+    if len(reference_parameters) > 0:
+        return reference_parameters[0].text, path_suffix
+    return None, path_suffix
+
+
+class SubscriptionBase:

```

Comment:
should this class be abstract?

Reply:
no, I think a Protocol definition is better. Added.